proposals/0003-gradle-kotlin-dsl.md

```diff
@@ -0,0 +1,55 @@
+---
+title: Use Kotlin DSL in gradle scripts
+author:
+- Dulmandakh Sukhbaatar
+date: 2018-08-11
+---
+
+# RFC0003: Use Kotlin DSL in gradle scripts
+
+## Summary
+
+Convert gradle scripts in ReactAndroid, and maybe RNTester, to Kotlin DSL.
+
+## Motivation
+
+Due to dynamic nature of Groovy language and lack of proper IDE support, its hard to make changes to gradle scripts because you have to run it to make sure that it works correctly. But running gradle scripts takes some time, and requires a developer to setup SDK and NDK, even for a trivial change.
+
+In 2016, Gradle introduced Kotlin programming language support in gradle scripts and it'll become stable in Gradle 4.10. Here is an excerpt from [announcement blog post](https://blog.gradle.org/kotlin-meets-gradle)
+
+> Because Kotlin is a statically-typed language with deep support in both IDEA and Eclipse, it could give Gradle users proper IDE support from auto-completion to refactoring and everything in-between. And because Kotlin is rich with features like first-class functions and extension methods, it could retain and improve on the best parts of writing Gradle build scripts—including a clean, declarative syntax and the ability to craft DSLs with ease.
+
+Thanks to static types and IDE support of Kotlin, IDE will offer following benefits, so that you don't have to run the gradle scripts.
+
+* auto-completion and content assist
+* quick documentation
+* navigation to source
+* refactoring and more
+
+Also, nowadays there are more Android developers who knows Kotlin than Groovy.
+
+## Detailed design
+
+Convert only gradle scripts in ReactAndroid, and maybe RNTester, to Kotlin DSL, so we don't introduce breaking changes to RN based projects.
```

Comment:
I think we should also consider scenarios where React Native is embedded into existing android apps. How will that work ?

Reply:
We may also want to consider the changes coming in React Fabric and Turbo Modules. In both cases, there would be code-gen solutions, possible in JS. How would that integrate with the build system? These Code gen solutions are not finalized, so I may not yet have specific examples.


Also, we may need to run annotation processors, like `ReactModuleInfoSpecProcessor` that we don't run today. Should we address that also ?